File: src/BmSDK.Host/dllmain.cpp
#include <Windows.h>

#include "Framework/offsets.h"
#include "Framework/detour_manager.h"

static void init_runtime() {
	offsets::BaseAddress = (uintptr_t)(GetModuleHandle(NULL));
	DetourRestoreAfterWith();
	DetourManager::RegisterEngineLoopPreInitDetour();
}

BOOL APIENTRY DllMain(HMODULE hModule, DWORD ul_reason_for_call, LPVOID lpReserved) {
	switch (ul_reason_for_call) {
		case DLL_PROCESS_ATTACH:
			init_runtime();
			break;
		case DLL_THREAD_ATTACH:
		case DLL_THREAD_DETACH:
		case DLL_PROCESS_DETACH:
			break;
	}
	return TRUE;
}

File: src/BmSDK.Host/DllMain.cs
using System.Runtime.InteropServices;
using Windows.Win32;
using Windows.Win32.Foundation;
using PInvokeDetours = Microsoft.Detours.PInvoke;

namespace BmSDK.Host;

static class DllMain
{
    const uint DLL_PROCESS_ATTACH = 1;

    [UnmanagedCallersOnly(EntryPoint = "DllMain")]
    static bool Main(nint hModule, uint reason, nint reserved)
    {
        if (reason == DLL_PROCESS_ATTACH)
        {
            InitRuntime();
        }

        return true;
    }

    static void InitRuntime()
    {
        Offsets.BaseAddress = (nuint)PInvoke.GetModuleHandle(default(PCWSTR)).Value;
        PInvokeDetours.DetourRestoreAfterWith();
        DetourManager.RegisterEngineLoopPreInitDetour();
    }
}
